hnn_core/network.py

```diff
@@ -488,3 +479,67 @@ def plot_spikes(self, ax=None, show=True):
         if show:
             plt.show()
         return ax.get_figure()
+
+    def build_in_neuron(self):
+        """
+        This function must be called before Network can be used for sims
+        """
+
+        from neuron import h
+        from .parallel import create_parallel_context, set_current_net
+
+        # make sure ParallelContext has been created (needed for joblibs)
+        create_parallel_context()
+
+        set_current_net(self)
+
+        self._gid_assign()
+        # Create a h.Vector() with size 1xself.N_t, zero'd
+        self.current = {
+            'L5Pyr_soma': h.Vector(self.N_t, 0),
+            'L2Pyr_soma': h.Vector(self.N_t, 0),
+        }
+        self._create_all_src()
+        self.state_init()
+        self._parnet_connect()
+        # set to record spikes
+        self.spiketimes = h.Vector()
+        self.spikegids = h.Vector()
+        self._record_spikes()
+        # position cells in 2D grid
+        self.move_cells_to_pos()
+
+    def clear_neuron_objects(self):
+        """
+        Clear up NEURON internal gid information.
+
+        Note: This function must be called from the context of the
+        Network instance that ran build_in_neuron. This is a bug or
+        peculiarity of NEURON. If this function is called from a different
+        context, then the next simulation will run very slow because nrniv
+        workers are still going for the old simulation. If pc.gid_clear is
+        called from the right context, then those workers can exit.
+        """
+        from .parallel import pc
+
+        pc.gid_clear()
+
+        # dereference cell and NetConn objects
+        for gid, cell in zip(self._gid_list, self.cells):
+            # only work on cells on this node
+            if pc.gid_exists(gid):
+                for name_src in ['L2Pyr', 'L2Basket', 'L5Pyr', 'L5Basket',
+                                 'extinput', 'extgauss', 'extpois', 'ev']:
+                    for nc in getattr(cell, 'ncfrom_%s' % name_src):
+                        if nc.valid():
+                            # delete NEURON cell object
+                            cell_obj1 = nc.precell(gid)
+                            if cell_obj1 is not None:
+                                del cell_obj1
```

Comment:
Even after gid_clear() these objects will remain internally referenced.
https://www.neuron.yale.edu/neuron/static/py_doc/modelspec/programmatic/network/parcon.html#ParallelContext.gid_clear

Reply:
uffff ... thanks for digging into this!